### PR TITLE
Adding the modified NB

### PR DESCRIPTION
**Pull request by noahweber1**
Concrete improvements from the previous notebook:

1. Complete restructuring and refactoring of the code.
2. Delete unecessary code
4. Seed set for complete reproducibility (such that we know when improvements are really improving and not due to randomness)
5. Decoupled the logic between data loading and processing
6. Wrote doc strings for most of the classes and functions
7. Exposed hyperparameters immediately at the beggining (for easier hydra refactoring)
8. Added more functionality (such as injectable/interchangable metric functions etc.)
9. Documented all of the main chapters and subchapters, with brief theoretical description behind stable diffusion, architecture tackled there, classifier-Free Diffusion Guidance and EMEA.

The main goal of this notebook is:

1. Set a major pre refactoring step such that we can easily move this to code base.
2. Set an easier entry point to newcomers.
3. Serve as the new benchmark for prototyping.
4. As the refactoring is ongoing all of the code here should be abstracted away from imports from the codebase.

### CODE REVIEW

File: notebooks/experiments/conditional_diffusion/dna_diff_baseline_conditional_UNET.ipynb [hunk 1]
```diff
@@ -0,0 +1,7119 @@
+{
```
Comment (SauravMaheshkar) [inline notebooks/experiments/conditional_diffusion/dna_diff_baseline_conditional_UNET.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    def compare_motif_list(df_motifs_a, df_motifs_b, motif_scoring_metric=motif_scoring_KL_divergence, plot_motif_probs=False):
</pre><ul><li>Are <code>df_motifs_a</code> and <code>df_motifs_b</code> going to be a pandas Series ?</li><li>We should add a <code>Callable</code> type hint to <code>motif_scoring_metric</code> .</li><li>We should add a <code>bool</code> type hint to <code>plot_motif_probs</code> .</li><li>We should add the <code>torch.Tensor</code> type hint to the function output.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinellolab/DNA-Diffusion/pull/60/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (SauravMaheshkar) [inline notebooks/experiments/conditional_diffusion/dna_diff_baseline_conditional_UNET.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    def metric_comparison_between_components(original_data, generated_data, x_label_plot, y_label_plot):
</pre><ul><li>We should add the <code>Dict</code> type hint to <code>original_data</code> and <code>generated_data</code>.</li><li>If I understand correctly <code>x_label_plot</code> and <code>y_label_plot</code> are strings (type: <code>str</code>) ? If yes we should add that.</li><li>We should add the <code>None</code> type hint to the function output.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinellolab/DNA-Diffusion/pull/60/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (SauravMaheshkar) [inline notebooks/experiments/conditional_diffusion/dna_diff_baseline_conditional_UNET.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    def one_hot_encode(seq, nucleotides, max_seq_len):
</pre><ul><li>Are the <code>seq</code> and <code>nucleotides</code> parameters, Lists or Strings ? The corresponding type hint should be added.</li><li>We should add the <code>int</code> type hint to <code>max_seq_len</code>.</li><li>We should add the <code>np.ndarray</code> type hint to the function output (viz. -&gt; <code>np.ndarray</code>)</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinellolab/DNA-Diffusion/pull/60/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (SauravMaheshkar) [inline notebooks/experiments/conditional_diffusion/dna_diff_baseline_conditional_UNET.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    def log(t, eps = 1e-20):
</pre><p>We should add the <code>torch.Tensor</code> type hint to the parameter <code>t</code> and the function output.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinellolab/DNA-Diffusion/pull/60/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (SauravMaheshkar) [inline notebooks/experiments/conditional_diffusion/dna_diff_baseline_conditional_UNET.ipynb:1]: <p>Same goes for all class methods except <code>update_average</code>. As they don't return any values we should add <code>None</code> to the method output.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinellolab/DNA-Diffusion/pull/60/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (SauravMaheshkar) [inline notebooks/experiments/conditional_diffusion/dna_diff_baseline_conditional_UNET.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.        def __init__(self, beta):
</pre><ul><li>We should ideally add class docstrings. The markdown comments above would be ideal.</li><li>We should add <code>float</code> type hint to the <code>beta</code> parameter.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinellolab/DNA-Diffusion/pull/60/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (SauravMaheshkar) [inline notebooks/experiments/conditional_diffusion/dna_diff_baseline_conditional_UNET.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.        def update_model_average(self, ma_model, current_model):
</pre><p>We should add <code>nn.Module</code> type hints to <code>ma_model</code> and <code>current_model</code> and <code>None</code> to the method output.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinellolab/DNA-Diffusion/pull/60/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (SauravMaheshkar) [inline notebooks/experiments/conditional_diffusion/dna_diff_baseline_conditional_UNET.ipynb:1]: <p>Great work on the abstraction 👍, type hints would be helpful.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinellolab/DNA-Diffusion/pull/60/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (SauravMaheshkar) [inline notebooks/experiments/conditional_diffusion/dna_diff_baseline_conditional_UNET.ipynb:1]: <p>Can we also do Garbage Collection and Empty cache after each step through the dataloader viz.</p><p> </p><pre class="ql-syntax" spellcheck="false">for epoch in range(...):
  for idx, sample in enumerate(dataloader):
  ...
  # ⭐️⭐️ Garbage Collection
  torch.cuda.empty_cache()
  _ = gc.collect()
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinellolab/DNA-Diffusion/pull/60/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='48'/>
